### PR TITLE
Use yieldDirectoryExclusivelyWithRelativePathname to allow filter by filename

### PR DESCRIPTION
**Pull request by alexander-schranz**
This way it is possible to use e.g.:

```php
vendor/bin/phpunit --filter="user_get_user_name_to_get_user_identifier.php.inc"
```

### CODE REVIEW

**Comment by TomasVotruba** [comment]
Does the number filter still works?

**Comment by alexander-schranz** [comment]
What you mean with number filter?

**Comment by samsonasik** [comment]
@alexander-schranz something like this:

```bash
vendor/bin/phpunit rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/AddArrayReturnDocTypeRectorTest.php --filter=#1
PHPUnit 9.5.20 #StandWithUkraine

.                                                                   1 / 1 (100%)

Time: 00:00.772, Memory: 64.50 MB

OK (1 test, 2 assertions)
```

I tried locally, and after apply your change, it seems cause empty result:

```bash
➜  rector-src git:(main) vendor/bin/phpunit rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/AddArrayReturnDocTypeRectorTest.php --filter=#1
PHPUnit 9.5.20 #StandWithUkraine

No tests executed!
```

so it make number filter **no longer working**.

**Comment by alexander-schranz** [comment]
Is that really required? I mean you normally don't know the number of the test but you see in your IDE the filename of the test. But we could still add if required `'#' . $counter` or something like that.

**Comment by samsonasik** [comment]
I usually doing that if found a numbered fixture error to just know the number of fixture, eg:

```diff
vendor/bin/phpunit rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/AddArrayReturnDocTypeRectorTest.php 
PHPUnit 9.5.20 #StandWithUkraine

................................F................................ 65 / 67 ( 97%)
..                                                                67 / 67 (100%)

Time: 00:04.125, Memory: 132.50 MB

There was 1 failure:

1) Rector\Tests\TypeDeclaration\Rector\ClassMethod\AddArrayReturnDocTypeRector\AddArrayReturnDocTypeRectorTest::test with data set #28 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/Fixture/fixture.php.inc
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
     /**
      * @return int[]
      */
-    public function getValues(): array
+    public function getValues()
```

above, I can see that there is error in fixture number `#28`, I don't need to copy fixture name, just number:

```diff
vendor/bin/phpunit rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/AddArrayReturnDocTypeRectorTest.php --filter=#28 
PHPUnit 9.5.20 #StandWithUkraine

F                                                                   1 / 1 (100%)

Time: 00:00.675, Memory: 60.50 MB

There was 1 failure:

1) Rector\Tests\TypeDeclaration\Rector\ClassMethod\AddArrayReturnDocTypeRector\AddArrayReturnDocTypeRectorTest::test with data set #28 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/Fixture/fixture.php.inc
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
     /**
      * @return int[]
      */
-    public function getValues(): array
+    public function getValues()
```

**Comment by alexander-schranz** [comment]
But it works the same for the new name instead of `#28` you will now get the filename `your_file.php.inc` and so you can use `--filter="your_file.php.inc"`. But if required we could prefix it with a counter.

**Comment by samsonasik** [comment]
I think that will make too long to type and can make typo as copy paste in the middle of long name. I think You can extends `AbstractRectorTestCase` and you can override it on your own use case.

File: packages/Testing/PHPUnit/AbstractRectorTestCase.php [hunk 1]
```diff
@@ -94,7 +94,12 @@ protected function tearDown(): void
      */
     protected function yieldFilesFromDirectory(string $directory, string $suffix = '*.php.inc'): Iterator
     {
-        return StaticFixtureFinder::yieldDirectoryExclusively($directory, $suffix);
+        $counter = 0;
+        foreach (StaticFixtureFinder::yieldDirectoryExclusivelyWithRelativePathname($directory, $suffix) as $name => $parameters) {
+            yield '#' . $counter . ' ' . $name => $parameters;
+
+            ++$counter;
+        }
```
Comment (samsonasik) [inline packages/Testing/PHPUnit/AbstractRectorTestCase.php:102]: it seems not working:

```bash
vendor/bin/phpunit rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/AddArrayReturnDocTypeRectorTest.php --filter=#1
PHPUnit 9.5.20 #StandWithUkraine

No tests executed!
```

Comment (alexander-schranz) [reply packages/Testing/PHPUnit/AbstractRectorTestCase.php:102]: Okay that is really strange seems like some strange phpunit internal behaviour how `#` is handled in the filter.

Still thx for your time and feedback.

**Pull request closed without merging**